File: C#7/Expression.cs
using System;

namespace ExpressionLib
{
    public class InvalidArgumentException : Exception
    {
        public InvalidArgumentException(string message) : base(message) { }
    }

    public class Expression
    {
        private double a;
        private double c;
        private double d;

        public Expression(double a, double c, double d)
        {
            this.a = a;
            this.c = c;
            this.d = d;
        }

        public double A => a;
        public double C => c;
        public double D => d;

        public void SetA(double value) => a = value;
        public void SetC(double value) => c = value;
        public void SetD(double value) => d = value;

        // Метод для логарифма
        private double CalculateLog(double x)
        {
            if (x <= 0)
                throw new InvalidArgumentException("Logarithm of non-positive number.");
            return Math.Log10(x);
        }

        // Основний метод обчислення виразу
        public double Calculate()
        {
            double denominator = a * a - 1;
            if (denominator == 0)
                throw new InvalidArgumentException("Division by zero.");

            double logPart = CalculateLog(d / 4.0);
            return (2 * c - logPart) / denominator;
        }
    }
}

File: C#6/C++7/Expression.cpp
#include "Expression.h"
#include <cmath>

Expression::Expression(double a, double c, double d) : a(a), c(c), d(d) {}

void Expression::setA(double value) { a = value; }
void Expression::setC(double value) { c = value; }
void Expression::setD(double value) { d = value; }

double Expression::getA() const { return a; }
double Expression::getC() const { return c; }
double Expression::getD() const { return d; }

double Expression::calculateLog(double x) const {
    if (x <= 0)
        throw InvalidArgumentException("Logarithm of non-positive number.");
    return log10(x);
}

double Expression::calculate() const {
    double denominator = a * a - 1;
    if (denominator == 0)
        throw InvalidArgumentException("Division by zero.");
    double logPart = calculateLog(d / 4.0);
    return (2 * c - logPart) / denominator;
}
